frame/preimage/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Preimage Pallet
+//!
+//! - [`Config`]
+//! - [`Call`]
+//!
+//! ## Overview
+//!
+//! The Preimage pallet allows for the users and the runtime to store the preimage
+//! of a hash on chain. This can be used by other pallets where storing and managing
+//! large byte-blobs.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_runtime::traits::{BadOrigin, Hash, One, Saturating};
+use sp_std::{convert::TryFrom, prelude::*};
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+	ensure,
+	pallet_prelude::Get,
+	traits::{Currency, ReservableCurrency},
+	weights::Pays,
+	BoundedVec,
+};
+use scale_info::TypeInfo;
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+pub use pallet::*;
+
+/// A type to note whether a preimage is owned by a user or the system.
+#[derive(Clone, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub enum RefCount<AccountId, Balance> {
+	/// This preimage is owned by a user who has the following deposit on hold.
+	User(AccountId, Balance),
+	/// This preimage is managed by the system with a number of reference counters.
+	System(u32),
+}
+
+/// The preimage metadata.
+#[derive(Clone, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub struct Preimage<BoundedVec, AccountId, Balance> {
+	// The preimage we are storing.
+	pub preimage: BoundedVec,
+	// A reference counter for who depends on this resource.
```

Comment:
```suggestion
	/// The preimage we are storing.
	pub preimage: BoundedVec,
	/// A reference counter for who depends on this resource.
```

and why is `BoundedVec` a generic?

Reply:
I think `MaxSize` would be a better type param to be passing. `Preimage` struct should know that it's a `BoundedVec<u8>`, and only need to be told of the size.